File: app/src/main/java/com/worldventures/dreamtrips/wallet/domain/converter/SmartCardFirmwareConverter.java
package com.worldventures.dreamtrips.wallet.domain.converter;

import com.worldventures.dreamtrips.modules.mapping.converter.Converter;
import com.worldventures.dreamtrips.wallet.domain.entity.ImmutableSmartCardFirmware;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCardFirmware;

import io.techery.janet.smartcard.model.FirmwareVersion;
import io.techery.mappery.MapperyContext;

class SmartCardFirmwareConverter implements Converter<FirmwareVersion, SmartCardFirmware> {

   @Override
   public SmartCardFirmware convert(MapperyContext mapperyContext, FirmwareVersion firmwareVersion) {
      return ImmutableSmartCardFirmware.builder()
            .firmwareVersion(firmwareVersion.appFirmwareVersion())
            .nrfBootloaderVersion(firmwareVersion.appBootloaderVersion())
            .internalAtmelVersion(firmwareVersion.internalAtmelVersion())
            .internalAtmelBootloaderVersion(firmwareVersion.internalAtmelBootloaderVersion())
            .externalAtmelVersion(firmwareVersion.externalAtmelVersion())
            .externalAtmelBootloaderVersion(firmwareVersion.externalAtmelBootloaderVersion())
            .build();
   }

   @Override
   public Class<FirmwareVersion> sourceClass() {
      return FirmwareVersion.class;
   }

   @Override
   public Class<SmartCardFirmware> targetClass() {
      return SmartCardFirmware.class;
   }
}

File: wallet/src/main/java/com/worldventures/wallet/domain/converter/SmartCardFirmwareConverter.kt
package com.worldventures.wallet.domain.converter

import com.worldventures.core.converter.Converter
import com.worldventures.wallet.domain.entity.SmartCardFirmware
import io.techery.janet.smartcard.model.FirmwareVersion
import io.techery.mappery.MapperyContext

internal class SmartCardFirmwareConverter : Converter<FirmwareVersion, SmartCardFirmware> {

   override fun convert(context: MapperyContext, source: FirmwareVersion): SmartCardFirmware {
      return SmartCardFirmware(
            nordicAppVersion = source.appFirmwareVersion(),
            nrfBootloaderVersion = source.appBootloaderVersion(),
            internalAtmelVersion = source.internalAtmelVersion(),
            internalAtmelBootloaderVersion = source.internalAtmelBootloaderVersion(),
            externalAtmelVersion = source.externalAtmelVersion(),
            externalAtmelBootloaderVersion = source.externalAtmelBootloaderVersion())
   }

   override fun sourceClass() = FirmwareVersion::class.java

   override fun targetClass() = SmartCardFirmware::class.java
}
